test/promql/tests/data_plane/gitops_sync_test.yaml

```diff
@@ -0,0 +1,84 @@
+evaluation_interval: 1m
+
+rule_files:
+  - prometheus.gitops_sync_alerts.yaml
+
+tests:
+  # Alerted cases:
+  - interval: 1m
+    input_series:
+
+      # GitOps service is experiencing application sync failure rate of less than 95%, so alert
+      - series: 'argocd_app_info{sync_status="Synced", namespace="gitops-service-argocd", pod="pod-1", source_cluster="cluster01"}'
+        values: '1+1x60'
+      - series: 'argocd_app_info{sync_status!="", namespace="gitops-service-argocd", pod="pod-1", source_cluster="cluster01"}'
+        values: '20+20x60'
+
+    alert_rule_test:
+      - eval_time: 60m
+        alertname: GitOpsApplicationSyncErrors
+        exp_alerts:
+          - exp_labels:
+              severity: warning
+              namespace: gitops-service-argocd
+              pod: pod-1
+              source_cluster: cluster01
+            exp_annotations:
+              summary: >-
+                The synced status percentage of all GitOps service applications is <95%
+              description: >-
+                The percentage total of all the Argo CD applications on cluster cluster01 having a sync
+                status of Synced is less than 95%.
+              runbook_url: https://gitlab.cee.redhat.com/rhtap/docs/sop/-/blob/main/gitops/deploy-from-git-to-k8s.md
+
+  - interval: 1m
+    input_series:
+
+      # GitOps Service experienced less than 95% synced rate in applications, should alert
+      - series: 'argocd_app_info{sync_status="Synced", namespace="gitops-service-argocd", pod="pod-1", source_cluster="cluster01"}'
+        values: '1+1x60'
```

Comment:
https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/

Reply:
Hi @yftacherzog , how about the sum?  Wouldn't it be the sum of all the apps present?  In the case I was thinking of, an app is created for each interval.  So it will be like this:

1 2 3 4 5.....61

But then, at the 51st interval, 20 new apps are added but because of the above, only 1 of the apps will be in Synced state.

```
Number In Synced:   1 2 3 4 5.....49 50 51 52 53.....61
           Total:   1 2 3 4 5.....49 50 70 71 72.....82
```

At that point, the 95% criteria is not met, so there should be an alert.

Let me know if my interpretation is way off :-)

Reply:
ok. I missed the part that the series created for the test was the sum rather than the raw data. It should be the raw data.

I also have some concerns about the validity of this query for this purpose. This is something that needs to be verified on a real environment to make sure that it behaves as expected.

My concern is like this: I think that the raw metric only gives value of 1 when the condition on the labels are satisfied and gives nothing when the conditions are not satisfied.

For example,  `argocd_app_info{sync_status="Synced", namespace="gitops-service-argocd", pod="pod-1", source_cluster="cluster01")}` will give 1 when it's in sync, but will not give 0 when it's out of sync.

I **suspect** that `argocd_app_info{health_status!="",sync_status=~"Synced"}` will keep returning the last value of the time series above, which is always 1.

So until the time series is cleaned out of the system (few days), this will keep returning 1.

This needs to be verified by forcing an app to be out of sync and checking if the sum changes.

I think the way around it will be to change the query to only take into account a time frame. e.g. `argocd_app_info{health_status!="",sync_status=~"Synced"}[5m]`, but that returns range vectors that you'd have to aggregate somehow, maybe with `last_over_time`.
e.g. `sum(last_over_time(argocd_app_info{health_status!="",sync_status=~"Synced"}[5m]))

All in all, the behavior needs to be verified on a proper environment and the unit test to be created in the end should be one that demonstrate how the data is present in Prometheus. I'd expect something like the following, but it needs to be verified:

```
      - series: 'argocd_app_info{sync_status="Synced", some_label="v1"}'
        values: '1 1 _ _ _ _ _ _'
      - series: 'argocd_app_info{sync_status="out-of-sync", some_label="v1"}'
        values: '_ _ 1 1 1 1 1 1'
      - series: 'argocd_app_info{sync_status="Synced", some_label="v2"}'
        values: '1 1 _ _ _ _ 1 1'
      - series: 'argocd_app_info{sync_status="out-of-sync", some_label="v2"}'
        values: '_ _ 1 1 1 1 _ _'
      - series: 'argocd_app_info{sync_status="Synced", some_label="v3"}'
        values: '1 1 1 1 1 1 1 1'
      - series: 'argocd_app_info{sync_status="out-of-sync", some_label="v3"}'
        values: '_ _ _ _ _ _ _ _'
```

Reply:
I've updated the tests and they run successfully